tianshou/data/buffer/her.py

```diff
@@ -0,0 +1,180 @@
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(size, **kwargs)
+        self.horizon = horizon
+        self.future_p = 1 - 1 / future_k
+        self.compute_reward_fn = compute_reward_fn
+        self._original_meta = Batch()
+        self._altered_indices = np.array([])
+
+    def _restore_cache(self) -> None:
+        """Write cached original meta back to `self._meta`.
+
+        It's called everytime before 'writing', 'sampling' or 'saving' the buffer.
+        """
+        if not hasattr(self, '_altered_indices'):
+            return
+
+        if self._altered_indices.size == 0:
+            return
+        self._meta[self._altered_indices] = self._original_meta
+        # Clean
+        del self._original_meta, self._altered_indices
+        self._original_meta = Batch()
+        self._altered_indices = np.array([])
+
+    def reset(self, keep_statistics: bool = False) -> None:
+        self._restore_cache()
+        return super().reset(keep_statistics)
+
+    def save_hdf5(self, path: str, compression: Optional[str] = None) -> None:
+        self._restore_cache()
+        return super().save_hdf5(path, compression)
+
+    def set_batch(self, batch: Batch) -> None:
+        self._restore_cache()
+        return super().set_batch(batch)
+
+    def update(self, buffer: Union["HERReplayBuffer", "ReplayBuffer"]) -> np.ndarray:
+        self._restore_cache()
+        return super().update(buffer)
+
+    def add(
+        self,
+        batch: Batch,
+        buffer_ids: Optional[Union[np.ndarray, List[int]]] = None
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        self._restore_cache()
+        return super().add(batch, buffer_ids)
+
+    def sample_indices(self, batch_size: int) -> np.ndarray:
+        """Get a random sample of index with size = batch_size.
+
+        Return all available indices in the buffer if batch_size is 0; return an \
+        empty numpy array if batch_size < 0 or no available index can be sampled. \
+        Additionally, some episodes of the sampled transitions will be re-written \
+        according to HER.
+        """
+        self._restore_cache()
+        indices = np.sort(super().sample_indices(batch_size=batch_size))
+        self.rewrite_transitions(indices)
+        return indices
+
+    def rewrite_transitions(self, indices: np.ndarray) -> None:
+        """Re-write the goal of some sampled transitions' episodes according to HER.
+
+        Currently applies only HER's 'future' strategy. The new goals will be written \
+        directly to the internal batch data temporarily and will be restored right \
+        before the next sampling or when using some of the buffer's method (e.g. \
+        `add`, `save_hdf5`, etc.). This is to make sure that n-step returns \
+        calculation etc., performs correctly without additional alteration.
+        """
+        if indices.size == 0:
+            return
+        # Construct episode trajectories
```

Comment:
I do think you should sort indices here. An example:
```
indices = [9, 7, 8, 6, 4, 5, 3, 1, 2, 0]
done = [0, 0, 1, 0, 0, 1, 0, 0, 1, 1]
```
after some processing with horizon=8, `unique_ep_open_indices = [7, 4, 1, 0]` and `unique_ep_close_indices=[ 3,  0, -1,  9]`, which makes no sense

Reply:
Previously I sort indices in `sample_indices` method (since it is the only one that calls `rewrite_transitions`) but yeah, I agree that sort indices should be inside `rewrite_transitions`.

fixed: https://github.com/thu-ml/tianshou/commit/63e65c087fa4c070861f12ba4d5fea130fc56740

Reply:
My bad. I didn't see L92 previously, but it also has an issue: if the replay buffer start index is not 0 (i.e., overwrite previous data point in a circular buffer), the sort approach will cause some bugs at some point. It should start with x where x is the closest index that is greater or equal to start index.

Reply:
You are right. This case somehow slipped my mind.
In `rewrite_transitions`, sorting indices serves as a way to group indices from the same episode together and order them chronologically. So to apply with the circular property, I'm thinking of sorting the indices, then shifting them so that the oldest index is at the front.
 
Will fix this later maybe today or tomorrow.

Reply:
Updated & add some test : https://github.com/thu-ml/tianshou/pull/753/commits/6ee7d08f414456ce2274e8ca0e6714e5cbe420c7